File: client/src/types/message.ts
export interface GameMessage {
  playerId: string;
  gameId:   string;
}

export interface MoveMessage extends GameMessage {
  type: "move";
  player: string;
  move: number;
}

export interface EventMessage extends GameMessage {
  type: string;
  payload?: any;
}

export const EventTypes = {
  START_GAME: 'start_game',
  WIN:        'win',
  LOSE:       'lose',
  DRAW:       'draw',
  MOVE:       'move',
}

export type EventType = typeof EventTypes[keyof typeof EventTypes] 

File: backend/pkg/message/message.go
package message

const (
	EventStartGame		= "start_game"
	EventWin					= "win"
	EventLose					= "lose"
	EventDraw					= "draw"
)

type GameMessage interface {
		GetPlayerId() string
		GetGameId()		string
}

type MoveMessage struct {
	Type			string	`json:"type"`
	PlayerId	string	`json:"playerId"`
	Player		string	`json:"player"`
	GameId		string	`json:"gameId"`
	Move			int			`json:"move"`
}

type EventMessage struct {
	Type			string			`json:"type"`
	PlayerId	string			`json:"playerId"`
	GameId		string			`json:"gameId"`
	Payload		interface{}	`json:"payload,omitempty"`
}

func (msg MoveMessage) GetGameId() string {
		return msg.GameId
}

func (msg MoveMessage) GetPlayerId() string {
		return msg.PlayerId
}

func NewStartGame(player, game, playerId string) EventMessage {
	return EventMessage{
		Type:			EventStartGame,
		PlayerId: playerId,
		GameId:		game,
		Payload:	player,
	}
}

func NewWinEvent(playerId string) EventMessage {
	return EventMessage{
		Type: EventWin,
		PlayerId: playerId,
	}
}

func NewLoseEvent(playerId string) EventMessage {
	return EventMessage{
		Type: EventLose,
		PlayerId: playerId,
	}
}

func NewDrawEvent(playerId string) EventMessage {
	return EventMessage{
		Type: EventDraw,
		PlayerId: playerId,
	}
}


